File: .claude/worktrees/agent-af1ba140/Qforge/measurement.py
# -*- coding: utf-8 -*-
# author: vinhpx
"""Quantum measurement operations."""
from __future__ import annotations
import numpy as np
import math
from Qforge._utils import _validate_qubit, _nq, _is_mps


def measure_all(wavefunction, n_samples: int):
    inds = np.random.choice(len(wavefunction.state), n_samples, p=wavefunction.probabilities())
    return np.unique(np.array(wavefunction.state[inds]), return_counts=True)

def _mps_prob0(wavefunction, n: int) -> float:
    """Get P(qubit n = |0>) from an MPS wavefunction."""
    if hasattr(wavefunction, '_mps') and wavefunction._mps is not None:
        return float(wavefunction._mps.measure_prob0(n))
    # Python fallback: full contraction
    from Qforge.mps import _contract_py
    amp = _contract_py(wavefunction._tensors)
    nq = len(wavefunction._tensors)
    prob_0 = sum(abs(amp[i])**2 for i in range(len(amp))
                 if not (i >> (nq - n - 1)) & 1)
    return float(prob_0)


def measure_one(wavefunction, n: int) -> np.ndarray:
    """Return probabilities [P(|0>), P(|1>)] for qubit n."""
    _validate_qubit(n, _nq(wavefunction))
    if wavefunction._sv is not None:
        prob_0 = round(wavefunction._sv.measure_one_prob0(n), 10)
        return np.array([prob_0, 1 - prob_0])
    if _is_mps(wavefunction):
        prob_0 = round(_mps_prob0(wavefunction, n), 10)
        return np.array([prob_0, 1 - prob_0])
    states = wavefunction.state
    amplitude = wavefunction.amplitude
    qubit_num = len(states[0])
    prob_0 = 0
    if n >= qubit_num or n < 0:
        raise ValueError("Index is out of range")
    for i in range(2**qubit_num):
        if states[i][n] == '0':
            prob_0 += abs(amplitude[i])**2
    prob_0 = round(prob_0, 10)
    return np.array([prob_0, 1 - prob_0])

def collapse_one(wavefunction, n: int) -> None:
    """Collapse qubit n into |0> or |1> via projective measurement."""
    _validate_qubit(n, _nq(wavefunction))
    if wavefunction._sv is not None:
        prob_0 = wavefunction._sv.measure_one_prob0(n)
        result_measure = np.random.choice(['0', '1'], 1, p=[prob_0, 1 - prob_0])[0]
        value = 0 if result_measure == '0' else 1
        wavefunction._sv.collapse_one(n, value)
        return
    if _is_mps(wavefunction):
        prob_0 = _mps_prob0(wavefunction, n)
        result_measure = np.random.choice(['0', '1'], 1, p=[prob_0, 1 - prob_0])[0]
        # Collapse: apply projector P0 or P1, then renormalize via amplitude setter
        from Qforge.mps import _contract_py, _svd_decompose
        amp = (_contract_py(wavefunction._tensors)
               if wavefunction._mps is None
               else wavefunction.amplitude)
        nq = _nq(wavefunction)
        mask = 1 << (nq - n - 1)
        if result_measure == '0':
            amp = np.where(np.arange(len(amp)) & mask == 0, amp, 0)
        else:
            amp = np.where(np.arange(len(amp)) & mask != 0, amp, 0)
        norm = np.linalg.norm(amp)
        if norm > 1e-14:
            amp /= norm
        wavefunction.amplitude = amp
        return
    states = wavefunction.state
    amplitude = wavefunction.amplitude
    qubit_num = len(states[0])
    prob_0 = 0
    new_amplitude = np.zeros(2**qubit_num, dtype = complex)
    if n >= qubit_num or n < 0:
        raise ValueError("Index is out of range")
    for i in range(2**qubit_num):
        if states[i][n] == '0':
            prob_0 += abs(amplitude[i])**2
    result_measure = np.random.choice(['0', '1'], 1, p = [prob_0, 1 - prob_0])[0]
    if result_measure == '0':
        for i in range(2**qubit_num):
            if states[i][n] == '0':
                new_amplitude[i] = amplitude[i]/math.sqrt(prob_0)
    elif result_measure == '1':
        for i in range(2**qubit_num):
            if states[i][n] == '1':
                new_amplitude[i] = amplitude[i]/math.sqrt(1 - prob_0)
    wavefunction.amplitude = new_amplitude

_PAULI_MAP = {'X': 0, 'Y': 1, 'Z': 2}

def pauli_expectation(wavefunction, qubit_idx: int, pauli_type: str) -> float:
    """Expectation value of Pauli operator ('X', 'Y', or 'Z') on qubit."""
    _validate_qubit(qubit_idx, _nq(wavefunction))
    if wavefunction._sv is not None:
        return wavefunction._sv.pauli_expectation(qubit_idx, _PAULI_MAP[pauli_type])
    if _is_mps(wavefunction):
        _PAULI_OPS = {
            'X': np.array([0, 1, 1, 0], dtype=complex),
            'Y': np.array([0, -1j, 1j, 0], dtype=complex),
            'Z': np.array([1, 0, 0, -1], dtype=complex),
        }
        op = _PAULI_OPS[pauli_type]
        if wavefunction._mps is not None:
            return float(wavefunction._mps.single_site_expectation(qubit_idx, op).real)
        from Qforge.dmrg import _single_site_expect_py
        return float(_single_site_expect_py(
            wavefunction._tensors, op.reshape(2, 2), qubit_idx).real)
    states = wavefunction.state
    amplitude = wavefunction.amplitude
    qubit_num = len(states[0])

    if qubit_idx >= qubit_num or qubit_idx < 0:
        raise ValueError("Index is out of range")

    expectation = 0.0

    if pauli_type == 'X':
        # <X> = sum of amplitude[i]*conj(amplitude[j]) where state[i] and state[j] differ at qubit_idx
        cut = 2**(qubit_num - qubit_idx - 1)
        for i in range(2**qubit_num):
            if states[i][qubit_idx] == '0':
                # <i|X|j> where j is i with flipped bit at qubit_idx
                j = i + cut
                expectation += 2 * np.real(np.conj(amplitude[i]) * amplitude[j])

    elif pauli_type == 'Y':
        # <Y> similar to X but with imaginary component
        cut = 2**(qubit_num - qubit_idx - 1)
        for i in range(2**qubit_num):
            if states[i][qubit_idx] == '0':
                j = i + cut
                expectation += 2 * np.real(-1j * np.conj(amplitude[i]) * amplitude[j])

    elif pauli_type == 'Z':
        # <Z> = P(|0>) - P(|1>)
        for i in range(2**qubit_num):
            prob = abs(amplitude[i])**2
            if states[i][qubit_idx] == '0':
                expectation += prob
            else:
                expectation -= prob

    return expectation

File: .claude/worktrees/agent-af1ba140/Qforge/cpp/src/measurement.cpp
#include "qforge/measurement.h"
#include <cmath>
#include <stdexcept>

namespace qforge { namespace measurement {

double measure_one_prob0(const StateVector& sv, int qubit) {
    const int n = sv.n_qubits();
    if (qubit < 0 || qubit >= n)
        throw std::out_of_range("Index is out of range");

    const size_t dim = sv.dim();
    const size_t mask = size_t(1) << (n - qubit - 1);
    const auto* amp = sv.data();
    double prob_0 = 0.0;

    #pragma omp parallel for reduction(+:prob_0) schedule(static) if(dim > 4096)
    for (size_t i = 0; i < dim; ++i) {
        if (!(i & mask)) {
            prob_0 += std::norm(amp[i]);
        }
    }
    return prob_0;
}

void collapse_one(StateVector& sv, int qubit, int value) {
    const int n = sv.n_qubits();
    if (qubit < 0 || qubit >= n)
        throw std::out_of_range("Index is out of range");
    if (value != 0 && value != 1)
        throw std::invalid_argument("value must be 0 or 1");

    const size_t dim = sv.dim();
    const size_t mask = size_t(1) << (n - qubit - 1);
    auto* amp = sv.data();

    // Compute probability of the target value
    double prob = 0.0;
    for (size_t i = 0; i < dim; ++i) {
        bool bit_is_one = (i & mask) != 0;
        if ((value == 0 && !bit_is_one) || (value == 1 && bit_is_one)) {
            prob += std::norm(amp[i]);
        }
    }

    if (prob < 1e-15)
        throw std::runtime_error("Cannot collapse to a state with zero probability");

    double norm_factor = 1.0 / std::sqrt(prob);

    // Normalize matching states, zero out non-matching
    #pragma omp parallel for schedule(static) if(dim > 4096)
    for (size_t i = 0; i < dim; ++i) {
        bool bit_is_one = (i & mask) != 0;
        if ((value == 0 && !bit_is_one) || (value == 1 && bit_is_one)) {
            amp[i] *= norm_factor;
        } else {
            amp[i] = {0.0, 0.0};
        }
    }
}

double pauli_expectation(const StateVector& sv, int qubit, int pauli_type) {
    const int n = sv.n_qubits();
    if (qubit < 0 || qubit >= n)
        throw std::out_of_range("Index is out of range");

    const size_t dim = sv.dim();
    const size_t mask = size_t(1) << (n - qubit - 1);
    const auto* amp = sv.data();
    double expectation = 0.0;

    if (pauli_type == 2) {
        // Z: <Z> = P(|0>) - P(|1>)
        #pragma omp parallel for reduction(+:expectation) schedule(static) if(dim > 4096)
        for (size_t i = 0; i < dim; ++i) {
            double prob = std::norm(amp[i]);
            if (!(i & mask))
                expectation += prob;
            else
                expectation -= prob;
        }
    } else if (pauli_type == 0) {
        // X: <X> = 2 * Re(sum conj(amp[i]) * amp[j]) for pairs where bit differs
        #pragma omp parallel for reduction(+:expectation) schedule(static) if(dim > 4096)
        for (size_t i = 0; i < dim; ++i) {
            if (!(i & mask)) {
                size_t j = i | mask;
                expectation += 2.0 * (amp[i].real() * amp[j].real() + amp[i].imag() * amp[j].imag());
            }
        }
    } else if (pauli_type == 1) {
        // Y: <Y> = 2 * Re(-i * conj(amp[i]) * amp[j])
        //        = 2 * (amp[i].real()*amp[j].imag() - amp[i].imag()*amp[j].real())
        #pragma omp parallel for reduction(+:expectation) schedule(static) if(dim > 4096)
        for (size_t i = 0; i < dim; ++i) {
            if (!(i & mask)) {
                size_t j = i | mask;
                expectation += 2.0 * (amp[i].real() * amp[j].imag() - amp[i].imag() * amp[j].real());
            }
        }
    } else {
        throw std::invalid_argument("pauli_type must be 0(X), 1(Y), or 2(Z)");
    }

    return expectation;
}

void probabilities(const StateVector& sv, double* out) {
    const size_t dim = sv.dim();
    const auto* amp = sv.data();

    #pragma omp parallel for schedule(static) if(dim > 4096)
    for (size_t i = 0; i < dim; ++i) {
        out[i] = std::norm(amp[i]);
    }
}

}} // namespace qforge::measurement
